Make go test ./... succeed by default

twpayne opened the pull request:
This PR makes

```console
$ go test ./...
```

succeed by default. Before, it would fail due to many unit tests actually being integration tests needing a running etcd server and the requirement for various linker flags to be present to set various test configuration variables.

See the individual commits for more details.

twpayne committed:
test: Use relative paths instead of testutils.CiliumRootDir

This removes the need for one extra linker flags for the tests to pass,
so the that extra linker flag is removed too.

Signed-off-by: Tom Payne <tom@isovalent.com>

twpayne committed:
treewide: Correct unit tests to integration tests

Before this commit, running the standard go command

    go test ./...

would fail because a number of unit tests were in fact integration
tests, typically requiring an etcd server to be running.

This PR adds an integration_test build flag to all such tests and
renames the unit-tests make target to integration-tests and sets this
flag.

With this commit "go test ./..." succeeds and runs a large subset of the
unit tests, completing in less than 15 seconds on a eight core machine
and a fully populated cache. "make integration-tests" in the same
configuration takes about nine minutes as it tests packages serially and
does not use or Go's test result cache.

Unfortunately, the bundling of many tests into gopkg.in/check.v1 suites
means that unit tests in many packages are coupled to integration
tests. This commit makes no attempt to unpick such coupling and so the
set of tests run by go test ./... is smaller than it could potentially
be. Improving this is left for a follow-up commit.

Signed-off-by: Tom Payne <tom@isovalent.com>

tklauser reviewed (changes requested):
There seems to be a mixture of `integration_test` (singular) and `integration_tests` (plural) used in the Go file build tags. I assume we want to consistently use `integration_tests` given that this is what's set in the `unit-tests` target in the `Makefile`.

joestringer commented on pkg/datapath/loader/datapath_sha256.go:
```diff
@@ -0,0 +1,5 @@
+package loader
```
This file is autogenerated, right? Should we avoid committing it to the tree?

tklauser replied:
In case we decide to add it to the tree, let's add a comment header saying:

```go
// Code generated by $(cmd). DO NOT EDIT
```

So it will be recognized by tooling and e.g., GitHub as auto-generated. Also see https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

twpayne replied:
For compatibility with the standard Go tooling we need to commit generated files to the tree. The `go` tool does not run any extra build steps by default, so if we don't commit the generated file, anyone checking out the code and running `go build` or `go test` will get an error about `DatapathSHA256` being undefined.

I've added the "Code generated..." header.

joestringer commented on pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go:
```diff
@@ -0,0 +1,46 @@
+// generate-cilium-datapath-sha256
```
Please add the SPDX + copyright lines here.

twpayne replied:
Done.

joestringer commented on pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go:
```diff
@@ -0,0 +1,46 @@
+// generate-cilium-datapath-sha256
+//
```
We also have the option to add this program under `tools/` rather than `pkg/datapath/loader/internal/...`, I think that'd be an easier place to find. At the same time this is likely a tool we write once and forget about, so I'm not too fussed if you prefer it to stay here.

twpayne replied:
Moved to `tools/`.

joestringer commented on Makefile:
```diff
@@ -124,7 +122,7 @@ define generate_k8s_protobuf
 endef
 
-build: $(SUBDIRS) ## Builds all the components for Cilium by executing make in the respective sub directories.
+build: pkg/datapath/loader/datapath_sha256.go $(SUBDIRS) ## Builds all the components for Cilium by executing make in the respective sub directories.
```
I *think* there might be a problem with parallel builds here. The `build` target says that it depends on the datapath sha and various subdirectories. So `make -jN` would try to build each of these dependencies independently, but if the scheduling of the datapath sha isn't quick enough then plausibly the daemon code would be compiled before the file is available (or updated), and the build could either fail (no such file) or encode the wrong datapath sha.

I think this file should be a dependency of the agent itself under `daemon/Makefile`?

twpayne replied:
Thanks, done.

joestringer commented on Makefile:
```diff
@@ -189,6 +189,7 @@ init-coverage: ## Initialize converage report for Cilium unit-tests.
 	$(QUIET) echo "mode: count" > coverage-all-tmp.out
 	$(QUIET) echo "mode: count" > coverage.out
 
+unit-tests: GO_TAGS_FLAGS+=integration_tests
```
I agree that some of these tests are integration tests, and tagging them as such is a good step forward.

What I like about this: Calling things out explicitly.

What I don't like about this particular line: Now we're explicitly defining the target `unit-tests` as integration tests, even though the majority of the tests are actually unit tests.

I'd like to see these to be part of a separate make target if we're going to call them out separately - maybe something like `make integration-tests` or `make go-integration-tests`?

At the same time, I don't want to lose coverage over these so if these particular tests are being run via a different target then let's make sure to grep around for where the `unit-tests` target is used, and update those locations to also call this target.

If we split this out, we can also move the `start-kvstores` / `stop-kvstores` logic over there, which will speed up this target. We can also update the docs so that we're not telling developers "to run the integration tests, run `make unit-tests`."

twpayne replied:
OK, I've renamed the `unit-tests` target to `integration-tests` and updated the documentation, which should resolve this.

joestringer commented on pkg/endpoint/endpoint_status_test.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// +build !privileged_tests
+// +build !privileged_tests,integration_tests
```
I don't think that most of these `pkg/endpoint` tests are actually integration tests.

I believe that due to tech debt around the code structure, they introduce dependencies on the BPF subsystem, but that's more of a code / compile issue rather than being relevant to the actual test execution.

The right solution is to deal with that technical debt. That's been a desirable item for a long time though, and it's not at all reasonable to require this from you in this PR. A middle ground might be to use a different tag, rather than `integration_tests` we have something more apt. But that probably makes it a bit more of a hassle to manage running those tests via the make targets... so if that's still pretty painful then I'm willing to accept this.

joestringer replied:
(I didn't look _too_ closely at these so I could also be off in my assessment here. If the comment seems off-base then feel free to ignore it).

twpayne replied:
OK if I fix the endpoint tests in a follow-up PR?

joestringer reviewed (changes requested):
Thanks for tackling this. I think there's just a couple of fairly minor bits I'd like to see changed but overall this seems like it will have a positive impact on the workflow both for general go familiarity, and also for developers on other platforms like osx who cannot run some of these more integration-style tests.

kkourt commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -51,7 +51,7 @@ var (
 	ep      = testutils.NewTestEndpoint()
 	hostEp  = testutils.NewTestHostEndpoint()
-	bpfDir  = filepath.Join(testutils.CiliumRootDir, "bpf")
+	bpfDir  = filepath.Join("..", "..", "..", "bpf")
```
Will this work regardless of where go test is invoked from?

That is, will it work both when running `go test ./...` from the top dir, `go test` inside pkg/datapath/loader, or any other option?

twpayne replied:
Yes, this will work. For any Go test, the working directory is the directory of the package (i.e. where the `_test.go` file is) so these relative paths work.

kkourt reviewed (approved):
Thanks!

LG from my side, assuming that invoking `go test` from any subdir would properly work (e.g., see comment) and that other concerns are also addressed.

tklauser commented on pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go:
```diff
@@ -0,0 +1,46 @@
+		}
+	}
+	_, err := fmt.Printf(""+
+		"package loader\n"+
```
As noted above on Joe's review, let's add a comment header saying:

```go
// Code generated by $(cmd). DO NOT EDIT
```

so the will be recognized by humans and tools as auto-generated. Also see https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

twpayne replied:
Done.

nebril commented on pkg/elf/elf_test.go:
```diff
@@ -36,7 +34,7 @@ type ELFTestSuite struct{}
 	_ = Suite(&ELFTestSuite{})
 
-	baseObjPath = filepath.Join(testutils.CiliumRootDir, "test", "bpf", "elf-demo.o")
+	baseObjPath = filepath.Join("..", "..", "test", "bpf", "elf-demo.o")
```
Same question as kkourt's above - will this work when run from within `pkg/elf`?

twpayne replied:
Yes.

nebril commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -28,13 +28,19 @@ both etcd and consul. To start the local instances, run:
 
-To run unit tests over the entire repository, run the following command in the
-project root directory:
+To run integration tests over the entire repository, run the following command
```
I know that this PR is a step forward in terms of naming the test suites correctly, but having `integration` here followed by `make unit-tests` would scream typo for me when reading the docs.

Would renaming the make target be possible, or is this out of scope for this PR?

Maybe we can create a `integration-tests` target as an alias for `unit-tests` so that the docs look good and we maintain backwards compatibility?

twpayne replied:
Done. I've renamed the `unit-tests` target to `integration-tests`.

jrajahalme commented on daemon/Makefile:
```diff
@@ -40,3 +41,9 @@ install-binary: install
 endif
+
+../pkg/datapath/loader/datapath_sha256.go: ../tools/generate-cilium-datapath-sha256/main.go $(BPF_SRCFILES_REL)
+	if [ "${GOARCH}" = "amd64"]; then \
```
Any reason this would not work on `arm64` as well?

joestringer replied:
Related, how does this not break the arm64 build?

joestringer replied:
:fearful: oh, because it's committed to the tree.

If we fail to update this datapath sha file, then it can break upgrades (basically if the datapath sha seems to be the same then there's no datapath changes therefore no need to replace the BPF programs). Let's be super careful here.

joestringer commented on Makefile:
```diff
@@ -177,19 +177,20 @@ ifeq ($(SKIP_KVSTORES),"false")
 	$(QUIET) echo "mode: count" > coverage.out
 
-unit-tests: start-kvstores ## Runs all unit-tests for Cilium.
+integration-tests: GO_TAGS_FLAGS+=integration_tests
+integration-tests: start-kvstores ## Runs all integration-tests for Cilium.
```
Now that the target is renamed, is this `integration-tests` target also running the unit tests? How do we know that both types of tests are continuing to be run as part of the CI? Bear with me, the answer to this might be simple but I couldn't tell from briefly perusing the changes.

What I had in mind was two targets, one for integration tests and one for unit tests, and these targets run the specific type of test you're asking for, and then we make doubly sure that we aren't omitting any of these tests from the testing infrastructure (ie Travis for unprivileged, Jenkins for privileged tests)

twpayne replied:
> Now that the target is renamed, is this `integration-tests` target also running the unit tests?

Yes. `go test ./...` runs all the unit tests. `go test -tags=integration_tests ./...` runs all the unit tests and the integration tests.

If we wanted separate unit and integration test targets we'd have to tag all the unit test files with a build tag so we can exclude them. I don't know if this is worth doing.

joestringer commented on pkg/envoy/envoy_version_sha256.go:
```diff
@@ -0,0 +1,11 @@
+// RequiredEnvoyVersionSHA is set during build
+// Running Envoy version will be checked against `RequiredEnvoyVersionSHA`.
+// By default cilium-agent will fail to start if there is a version mismatch.
+const RequiredEnvoyVersionSHA = "3ae9ae3ca32eb30816c9e6a655dcb36e7ba4183c8c0c93ad58b72ad6c064aa0b"
```
If we're committing these to the tree now, does this cause weirdness with git any time we build? Should we add these files to .gitignore?

Also, do we need to update these SHAs for each release or anything like that? Or are these just dummy values? In which case should we make them clearly dummy values using a magic pattern?

twpayne replied:
These files should not be in `.gitignore` as they are needed for Cilium to build from a clean checkout.

> Also, do we need to update these SHAs for each release or anything like that?

The SHA256s should be updated automatically whenever you run `make`. As the generated files are checked into git, any changes to them will be visible in `git diff` and should be checked in alongside the commit that causes the SHA256 to change.

joestringer replied:
> These files should not be in .gitignore as they are needed for Cilium to build from a clean checkout.

Would adding them to the tree + adding to .gitignore solve this? I understand that we need it available to be able to build, but it sounds like if the build is working correctly then any time we change bpf/ code, this file will change; but we don't really want to track the changes to the file.

> As the generated files are checked into git, any changes to them will be visible in git diff and should be checked in alongside the commit that causes the SHA256 to change.

We used to have a system like this in the past, it's a pain to maintain. This means if there are ever parallel BPF changes in flight, we are forced to rebase BPF PRs purely to satisfy this constraint :/

joestringer replied:
^^ I think we should raise awareness of this proposal for this last point and see if we can come up with a better solution.

joestringer commented on pkg/datapath/loader/hash_test.go:
```diff
@@ -44,6 +46,7 @@ func (s *LoaderTestSuite) TesthashDatapath(c *C) {
 
 	// Ensure we get the same base hash when config is removed via Reset()
 	h.Reset()
+	io.WriteString(h, DatapathSHA256)
```
Why is this necessary? Does the test break? The idea behind the "DatapathHash" object is that if you reset it, it goes back to the default SHA so if this test now needs to write the sha inside, then maybe the object isn't resetting correctly (in which case this could indicate a runtime bug).

twpayne replied:
This is necessary because [`newDatapathHash()` seeds the hash with `DatapathSHA256`](https://github.com/cilium/cilium/blob/master/pkg/datapath/loader/hash.go#L43).

I'm a bit suspicious of the datapath hashing code, but I consider it a separate issue to making `go test ./...` to succeed by default, so I've done the minimum changes to get the tests to pass in this PR.

joestringer replied:
:thinking: newDatapathHash() always seeded the hash with DatapathSHA256. 

When I look a little closer, it seems like there is no `h.Reset()` outside of unit tests, so the comment on line 45->47 is actually asserting behaviour that is not implemented and not used. That's part of where this discrepancy came from.  Honestly it looks like we should just remove these 4 lines of test + comment.

I think that maybe at some point, `Reset()` was expected to be implemented and used for resumeable hashes (to ensure that the hash was always seeded with the base datapath sha), but in the end we landed with `sumEndpoint()` which makes a copy of the hash every time, so it's actually not necessary.

twpayne committed:
test: Remove unused ldflags

The ldflag setting
github.com/cilium/cilium/pkg/kvstore.consulDummyConfigFile set the
existing value.

github.com/cilium/cilium/pkg/logging.DefaultLogLevelStr is no longer
present.

Signed-off-by: Tom Payne <tom@isovalent.com>

joestringer commented on test/Makefile:
```diff
@@ -25,13 +25,13 @@ build:
 
 run:
-	ginkgo --focus "Runtime" -v -- --cilium.provision=$(PROVISION) --cilium.registryCredentials=$(REGISTRY_CREDENTIALS)
+	ginkgo --focus "Runtime" --tags integration_tests -v -- --cilium.provision=$(PROVISION) --cilium.registryCredentials=$(REGISTRY_CREDENTIALS)
```
Do we need to do this for the `build` target as well?

Upstream docs say:

```
Build flags:
    -tags
        Set -tags to pass in build tags down to the compilation step.
```

twpayne replied:
No, we don't need to add `-tags=integration_tests` to the build target as the tag is only used in tests.

twpayne replied:
> One minor thing to check on the `make -C test build` target.

Ah, I see what you mean now. AFAICT the resulting `test.test` binary is not used anywhere, but I've updated the target to add the `integration_tests` tag.

joestringer commented on test/test_test.go:
```diff
@@ -1,5 +1,7 @@
-// Copyright 2017 Authors of Cilium
+// Copyright 2017-2021 Authors of Cilium
+
+// +build integration_tests
```
It'd be nice if we could figure out a way to allow both `go test ./...` and also `ginkgo -focus ...` without adding this extra `-tags integration_tests` boilerplate argument for every time a developer wants to run ginkgo :thinking: 

At the same time, I personally have macros wrapped around the ginkgo invocations anyway so it doesn't directly affect me. And I'd probably rather that unit-testing is the easiest type of testing in order to encourage developers to write more of them.

twpayne replied:
> It'd be nice if we could figure out a way to allow both `go test ./...` and also `ginkgo -focus ...` without adding this extra `-tags integration_tests` boilerplate argument for every time a developer wants to run ginkgo 🤔

Agree, I'd prefer this as well, but I can't find a mechanism to do it, short of providing our own `ginkgo` binary.

joestringer reviewed (approved):
One minor thing to check on the `make -C test build` target. Not sure why you're hitting so much turmoil in the CI, is the tree in general this unstable? :thinking:

gandro reviewed (approved):
:rocket:
